contracts/test/PDPService.t.sol

```diff
@@ -313,6 +316,263 @@ contract PDPServiceProofSetMutateTest is Test {
+    }
+
+    function testLateProofAccepted() public {
+        uint leafCount = 10;
+        (uint256 setId, bytes32[][] memory tree) = makeProofSetWithOneRoot(10);
+
+        // Advance chain short of challenge epoch
+        uint256 challengeEpoch = pdpService.getNextChallengeEpoch(setId);
+        vm.roll(challengeEpoch + 100);
+
+        // Build a proof.
+        PDPService.Proof[] memory proofs = buildProofsForSingleton(setId, 1, tree, leafCount);
+
+        // Submit proof.
+        pdpService.provePossession(setId, proofs);
+        assertEq(pdpService.getNextChallengeEpoch(setId), block.number + challengeFinalityDelay);
+        assertEq(pdpService.getNextChallengeEpoch(setId), block.number + challengeFinalityDelay);
+    }
+
+    function testEarlyProofRejected() public {
+        uint leafCount = 10;
+        (uint256 setId, bytes32[][] memory tree) = makeProofSetWithOneRoot(10);
+
+        // Advance chain short of challenge epoch
+        uint256 challengeEpoch = pdpService.getNextChallengeEpoch(setId);
+        vm.roll(challengeEpoch - 1);
+
+        // Build a proof.
+        PDPService.Proof[] memory proofs = buildProofsForSingleton(setId, 1, tree, leafCount);
+
+        // Submit proof.
+        vm.expectRevert();
+        pdpService.provePossession(setId, proofs);
+    }
+
+    function testEmptyProofRejected() public {
+        uint256 setId = pdpService.createProofSet();
+        PDPService.Proof[] memory emptyProof = new PDPService.Proof[](0);
+
+        // Rejected with no roots
+        vm.expectRevert();
+        pdpService.provePossession(setId, emptyProof);
+
+        addOneRoot(setId, 10);
+
+        // Rejected with a root
+        vm.expectRevert();
+        pdpService.provePossession(setId, emptyProof);
+    }
+
+    function testBadChallengeRejected() public {
+        uint leafCount = 10;
+        (uint256 setId, bytes32[][] memory tree) = makeProofSetWithOneRoot(10);
+
+        // Make a proof that's good for this challenge epoch.
+        uint256 challengeEpoch = pdpService.getNextChallengeEpoch(setId);
+        vm.roll(challengeEpoch);
+        PDPService.Proof[] memory proofs = buildProofsForSingleton(setId, 1, tree, leafCount);
+
+        // Submit proof successfully, advancing the proof set to a new challenge epoch.
+        pdpService.provePossession(setId, proofs);
+
+        uint nextChallengeEpoch = pdpService.getNextChallengeEpoch(setId);
+        assertNotEq(nextChallengeEpoch, challengeEpoch);
+        vm.roll(nextChallengeEpoch);
+        
+        // The proof for the old challenge epoch should no longer be valid.
+        vm.expectRevert();
+        pdpService.provePossession(setId, proofs);
+    }
+
+    function testBadRootsRejected() public {
+        uint[] memory leafCounts = new uint[](2);
+        // Note: either co-prime leaf counts or a challenge count > 1 are required for this test to demonstrate the failing proof.
```

Comment:
I'm not sure about this but my best understanding is the co prime condition needs to exist over leaf count1 and (leaf count1 + leaf count2)